File: plugins/eclipse/online.xprei.ide.eclipse/src/online/xprei/ide/eclipse/WebviewResources.java
package online.xprei.ide.eclipse;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.file.Files;

/**
 * Extracts the plugin's bundled resources (the shared {@code webview/}, and
 * the esbuild-bundled {@code sidecar.cjs} — both physically copied into this
 * bundle at build time by the child pom.xml's maven-resources-plugin
 * executions, listed in build.properties's bin.includes) to real files on
 * disk. SWT's {@code Browser.setUrl()} needs a real {@code file://} URL, and
 * {@code ProcessBuilder} needs a real path to spawn {@code node <path>}
 * against.
 */
public final class WebviewResources {
    private static final String[] WEBVIEW_FILES = {
        "index.html", "chat.js", "chat.css", "bridge.js", "theme-fallback.css", "icon.png", "icon.svg",
    };

    private static File extractedWebviewDir;
    private static File extractedSidecarPath;

    private WebviewResources() {}

    public static synchronized File extractWebviewDir() throws IOException {
        if (extractedWebviewDir != null) return extractedWebviewDir;
        File dir = Files.createTempDirectory("xprei-webview").toFile();
        for (String name : WEBVIEW_FILES) {
            try (InputStream in = WebviewResources.class.getClassLoader().getResourceAsStream("webview/" + name)) {
                if (in == null) continue; // best-effort; a missing icon shouldn't break chat
                copy(in, new File(dir, name));
            }
        }
        extractedWebviewDir = dir;
        return dir;
    }

    public static synchronized File extractSidecar() throws IOException {
        if (extractedSidecarPath != null) return extractedSidecarPath;
        File dir = Files.createTempDirectory("xprei-sidecar").toFile();
        File target = new File(dir, "sidecar.cjs");
        try (InputStream in = WebviewResources.class.getClassLoader().getResourceAsStream("sidecar.cjs")) {
            if (in == null) {
                throw new IOException(
                    "xpreiIDE: sidecar.cjs missing from plugin resources. Was `npm run build:sidecar -w "
                        + "@xprei/core` run before the Maven build? (The child pom's copy-sidecar execution "
                        + "should do this automatically.)");
            }
            copy(in, target);
        }
        extractedSidecarPath = target;
        return target;
    }

    private static void copy(InputStream in, File dest) throws IOException {
        try (OutputStream out = Files.newOutputStream(dest.toPath())) {
            byte[] buf = new byte[8192];
            int n;
            while ((n = in.read(buf)) > 0) out.write(buf, 0, n);
        }
    }
}

File: plugins/intellij/src/main/kotlin/online/xprei/ide/WebviewResources.kt
package online.xprei.ide

import java.io.File
import java.nio.file.Files

// Extracts the plugin's bundled classpath resources (the shared webview/, and
// the esbuild-bundled sidecar.cjs — both wired onto the classpath by
// build.gradle.kts's sourceSets.main.resources.srcDir) to real files on disk.
// JBCefBrowser.loadURL() needs an actual file:// URL (a jar:// URL won't
// resolve the webview's own relative asset links), and ProcessBuilder needs a
// real path to spawn `node <path>` against.
object WebviewResources {
    private val WEBVIEW_FILES = listOf(
        "index.html", "chat.js", "chat.css", "bridge.js", "theme-fallback.css",
        "icon.png", "icon.svg",
    )

    @Volatile
    private var extractedWebviewDir: File? = null

    @Volatile
    private var extractedSidecarPath: File? = null

    @Synchronized
    fun extractWebviewDir(): File {
        extractedWebviewDir?.let { return it }
        val dir = Files.createTempDirectory("xprei-webview").toFile()
        for (name in WEBVIEW_FILES) {
            val stream = javaClass.classLoader.getResourceAsStream(name) ?: continue
            stream.use { input -> File(dir, name).outputStream().use { input.copyTo(it) } }
        }
        extractedWebviewDir = dir
        return dir
    }

    @Synchronized
    fun extractSidecar(): File {
        extractedSidecarPath?.let { return it }
        val dir = Files.createTempDirectory("xprei-sidecar").toFile()
        val target = File(dir, "sidecar.cjs")
        val stream = javaClass.classLoader.getResourceAsStream("sidecar.cjs")
            ?: error(
                "xpreiIDE: sidecar.cjs missing from plugin resources. " +
                    "Was `npm run build:sidecar -w @xprei/core` run before the Gradle build? " +
                    "(build.gradle.kts's buildSidecar task should do this automatically.)",
            )
        stream.use { input -> target.outputStream().use { input.copyTo(it) } }
        extractedSidecarPath = target
        return target
    }
}
